python/lsst/rubintv/production/watchers.py

```diff
@@ -0,0 +1,226 @@
+
+    When a new expRecord is found, it writes it out to a file so that it
+    can be found by a FileWatcher, so that we can poll for new data without
+    hammering the main repo with 201x ButlerWatchers.
+
+    Only one of these should be instantiated per-location.
+
+    Parameters
+    ----------
+    butler : `lsst.daf.butler.Butler`
+        The butler.
+    locationConfig : `lsst.rubintv.production.utils.LocationConfig`
+        The location config.
+    dataProducts : `str` or `list` [`str`]
+        The data products to watch for.
+    doRaise : `bool`, optional
+        Raise exceptions or log them as warnings?
+    """
+    # look for new images every ``cadence`` seconds
+    cadence = 1
+    # upload heartbeat every n seconds
+    HEARTBEAT_UPLOAD_PERIOD = 30
+    # consider service 'dead' if this time exceeded between heartbeats
+    HEARTBEAT_FLATLINE_PERIOD = 120
+
+    def __init__(self, locationConfig, butler, dataProducts, doRaise=False):
+        self.locationConfig = locationConfig
+        self.butler = butler
+        self.dataProducts = list(ensure_iterable(dataProducts))  # must call list or we get a generator back
+        self.doRaise = doRaise
+        self.log = _LOG.getChild("butlerWatcher")
+
+    def _getLatestExpRecords(self):
+        """Get the most recent expRecords from the butler.
+
+        Get the most recent expRecord for all the dataset types. These are
+        written to files for the FileWatchers to pick up.
+
+        Returns
+        -------
+        expRecords : `dict` [`str`, `lsst.daf.butler.DimensionRecord`]
+            A dict of the most recent exposure records, keyed by dataProduct.
+        """
+        expRecordDict = {}
+
+        # this where clause is no longer necessary for speed now that we are
+        # using order_by and limit, but it now functions to restrict data to
+        # being inside the range where exposures could ever realistically be
+        # taken, as some simulated data is "taken" in the year 3000. We must
+        # also restrict the exposure.id specifically, as the day_obs seems to
+        # be set to this millenium, but the expIds are like 3YYYMMDDNNNNN.
+        where = ('exposure.day_obs>20200101 and exposure.day_obs<21000101'
+                 ' and exposure.id<2100010100001')
+
+        for product in self.dataProducts:
+            # NB if you list multiple products for datasets= then it will only
+            # give expRecords for which all those products exist, so these must
+            # be done as separate queries
+            records = self.butler.registry.queryDimensionRecords("exposure",
+                                                                 datasets=product,
+                                                                 where=where)
+            records.order_by('-exposure.id')  # the minus means descending ordering
+            records.limit(1)
+            expRecordDict[product] = list(records)[0]
```

Comment:
Do you need to worry here about the query not returning anything?

Reply:
I wouldn't have thought so, but as it turns out, yes, yes I do 🙂

Reply:
(this is because of the way TS8 does its exposure ids, meaning they're a factor of 10 larger than on the summit) Also I need to protect for dataProducts which don't exist, so double-yes!